sklearn/cluster/tests/test_hierarchical.py

```diff
@@ -527,13 +528,141 @@ def test_agg_n_clusters():
+
+    rng = np.random.RandomState(0)
+    mask = np.ones([10, 10], dtype=np.bool)
+    n_samples = 100
+    X = rng.randn(n_samples, 50)
+    connectivity = grid_to_graph(*mask.shape)
+    # test when distance threshold is set to 10
+    distance_threshold = 10
+    for linkage in ("ward", "complete", "average"):
+        clustering = AgglomerativeClustering(
+            n_clusters=None,
+            distance_threshold=distance_threshold,
+            connectivity=connectivity, linkage=linkage)
+        clustering.fit(X)
+        clusters_produced = clustering.labels_
+        num_clusters_produced = len(np.unique(clustering.labels_))
+        # test if the clusters produced match the point in the linkage tree
+        # where the distance exceeds the threshold
+        tree_builder = _TREE_BUILDERS[linkage]
+        children, n_components, n_leaves, parent, distances = \
+            tree_builder(X, connectivity=connectivity, n_clusters=None,
+                         return_distance=True)
+        clusters_at_threshold = _hc_cut(n_clusters=None, children=children,
+                                        n_leaves=n_leaves,
+                                        distance_threshold=distance_threshold,
+                                        distances=distances)
+        num_clusters_at_threshold = np.count_nonzero(
+            distances >= distance_threshold) + 1
+        # test number of clusters produced
+        assert_true(num_clusters_at_threshold == num_clusters_produced)
+        # test clusters produced
+        assert_true(np.array_equiv(clusters_produced, clusters_at_threshold))
+
+        # test if cluster number is different without connectivity
+        clustering.connectivity = None
+        clustering.fit(X)
+        clusters_produced = clustering.labels_
+        num_clusters_produced = len(np.unique(clustering.labels_))
+        # test if the clusters produced match the point in the linkage tree
+        # where the distance exceeds the threshold
+        tree_builder = _TREE_BUILDERS[linkage]
+        children, n_components, n_leaves, parent, distances = \
+            tree_builder(X, connectivity=None, n_clusters=None,
+                         return_distance=True)
+        clusters_at_threshold = _hc_cut(n_clusters=None, children=children,
+                                        n_leaves=n_leaves,
+                                        distance_threshold=distance_threshold,
+                                        distances=distances)
+        num_clusters_at_threshold = np.count_nonzero(
+            distances >= distance_threshold) + 1
+        # test number of clusters produced
+        assert_true(num_clusters_at_threshold == num_clusters_produced)
+        # test clusters produced
+        assert_true(np.array_equiv(clusters_produced, clusters_at_threshold))
+
+
+def test_agglomerative_clustering_with_distance_threshold_edge_case():
+    # test boundary case of distance_threshold matching the distance
```

Comment:
This test is okay, but I think it would be simpler to read and understand if you just do something like:

```
    X = [[0], [1]]
    for threshold, y_true in [(0.5, ...), (1.0, ...), (1.5, ...)]:
        # TODO: perhaps also loop over linkage
        clusterer = AgglomerativeClustering(n_clusters=None, distance_threshold=threshold)
        y_pred = clusterer.fit_predict(X)
        assert_equal(1, adjusted_rand_score(y_true, y_pred))
```